dbms/src/Interpreters/Join.cpp

```diff
@@ -378,6 +379,15 @@ std::shared_ptr<Join> Join::createRestoreJoin(size_t max_bytes_before_external_j
         flag_mapped_entry_helper_name,
         probe_cache_column_threshold,
         is_test);
+    /// init output names after finalize, the restored join don't need to finalize
+    ret->output_columns_after_finalize = output_columns_after_finalize;
+    ret->output_column_names_set_after_finalize = output_column_names_set_after_finalize;
+    ret->output_columns_names_set_for_other_condition_after_finalize
+        = output_columns_names_set_for_other_condition_after_finalize;
+    ret->required_columns = required_columns;
+    ret->output_block_after_finalize = output_block_after_finalize;
```

Comment:
why not using `output_block` directly?

Reply:
`output_block` is already inited in the constructor of `Join`, here we just assign some finalized releated field because the `finalize` method of the newly constructed `Join` is not called.